src/win32/Watcher.cpp

```diff
@@ -0,0 +1,276 @@
+    if (errorCode == ERROR_NOTIFY_ENUM_DIR) {
+      setError("Buffer filled up and service needs a restart");
+    } else if (errorCode == ERROR_INVALID_PARAMETER) {
+      // resize the buffers because we're over the network, 64kb is the max buffer size for networked transmission
+      resizeBuffers(64 * 1024);
+
+      if (!loop()) {
+        setError("failed resizing buffers for network traffic");
+      }
+    } else {
+      setError("Service shutdown unexpectedly");
+    }
+    return;
+  }
+
+  std::swap(mWriteBuffer, mReadBuffer);
+  loop();
+  handleEvents();
+}
+
+void Watcher::handleEvents()
+{
+  BYTE *base = mReadBuffer.data();
+  while (true) {
+    PFILE_NOTIFY_INFORMATION info = (PFILE_NOTIFY_INFORMATION)base;
+    std::wstring fileName = getWStringFileName(info->FileName, info->FileNameLength);
+
+    if (info->Action == FILE_ACTION_RENAMED_OLD_NAME) {
+      if (info->NextEntryOffset != 0) {
+        base += info->NextEntryOffset;
+        info = (PFILE_NOTIFY_INFORMATION)base;
+        if (info->Action == FILE_ACTION_RENAMED_NEW_NAME) {
+          std::wstring fileNameNew = getWStringFileName(info->FileName, info->FileNameLength);
+
+          mQueue->enqueue(
+              RENAMED,
+              getUTF8Directory(fileName),
+              getUTF8FileName(fileName),
+              getUTF8FileName(fileNameNew)
+          );
+        } else {
+          mQueue->enqueue(DELETED, getUTF8Directory(fileName), getUTF8FileName(fileName));
+          continue;
+        }
+      } else {
+        mQueue->enqueue(DELETED, getUTF8Directory(fileName), getUTF8FileName(fileName));
+        break;
+      }
+    }
+
+    switch (info->Action) {
+    case FILE_ACTION_ADDED:
+    case FILE_ACTION_RENAMED_NEW_NAME: // in the case we just receive a new name and no old name in the buffer
+      mQueue->enqueue(CREATED, getUTF8Directory(fileName), getUTF8FileName(fileName));
+      break;
+    case FILE_ACTION_REMOVED:
+      mQueue->enqueue(DELETED, getUTF8Directory(fileName), getUTF8FileName(fileName));
+      break;
+    case FILE_ACTION_MODIFIED:
+    default:
+      mQueue->enqueue(MODIFIED, getUTF8Directory(fileName), getUTF8FileName(fileName));
+    };
+
+    if (info->NextEntryOffset == 0) {
+      break;
+    }
+    base += info->NextEntryOffset;
+  }
+}
+
+void Watcher::start()
+{
+  mRunner = std::thread([this]{
+    // mRunning is set to false in the d'tor
+    mRunning = true;
+    run();
+  });
+
+  if (!mRunner.joinable()) {
+    mRunning = false;
+    return;
+  }
+
+  QueueUserAPC([](__in ULONG_PTR self) {
+    auto watcher = reinterpret_cast<Watcher*>(self);
+    watcher->mHasStartedSemaphore.signal();
+    watcher->loop();
+  } , mRunner.native_handle(), (ULONG_PTR)this);
+
+  if (!mHasStartedSemaphore.waitFor(std::chrono::seconds(10))) {
+    setError("Watcher is not started");
+  }
+}
+
+void Watcher::stop()
+{
+  mRunning = false;
+  QueueUserAPC([](__in ULONG_PTR) {}, mRunner.native_handle(), (ULONG_PTR)this);
```

Comment:
Can we add a comment here, it would be a lot faster for future contributors to recognize that this is to wake up the SleepEx above so that the watcher's loop exits cleanly.